src/test/java/com/sap/sgs/phosphor/fosstars/data/github/GitHubDataFetcherTest.java

```diff
@@ -79,4 +90,227 @@ public void testShouldUpdate() throws IOException {
+            fetcher.localRepositoryFor(project);
+          } catch (IOException e) {
+            throw new IllegalStateException(e);
+          }
+          latch.countDown();
+        });
+      }
+      assertTrue(latch.await(5, TimeUnit.SECONDS));
+
+      checkLocalRepository(project, 1);
+    } finally {
+      executorService.shutdown();
+    }
+  }
+
+  @Test
+  public void testMultiThreadCleanUpTest() throws IOException, InterruptedException {
+    ExecutorService executorService = Executors.newFixedThreadPool(10);
+    GitHubProject project = new GitHubProject("test", "project");
+    Path projectDir = directoryFor(project);
+
+    try (
+        Repository repository = FileRepositoryBuilder.create(projectDir.resolve(".git").toFile())) {
+      repository.create();
+      fetcher.addRepositoryInfoForTesting(project, projectDir);
+      testLocalRepositoryFor(project, 1);
+
+      int numberOfThreads = 25;
+      CountDownLatch latch = new CountDownLatch(numberOfThreads);
+      for (int i = 0; i < numberOfThreads; i++) {
+        executorService.submit(() -> {
+          try {
+            runCleanupFor(project);
+          } catch (IOException e) {
+            throw new IllegalStateException("Clean up exception");
+          }
+          latch.countDown();
+        });
+      }
+      assertTrue(latch.await(5, TimeUnit.SECONDS));
+
+      checkCleanUp(project, 0);
+    } finally {
+      executorService.shutdown();
+    }
+  }
+
+  @Test
+  public void testMultipleInstanceFetcher() throws IOException {
+    // Create the first project with first GitHub mock instance.
+    GitHubProject firstProject = new GitHubProject("test", "project1");
+    Path project1Dir = directoryFor(firstProject);
+    GitHub github = mock(GitHub.class);
+    try (Repository repository =
+        FileRepositoryBuilder.create(project1Dir.resolve(".git").toFile())) {
+      repository.create();
+
+      fetcher = new TestGitHubDataFetcher(github);
+      fetcher.addRepositoryInfoForTesting(firstProject, project1Dir);
+      testLocalRepositoryFor(firstProject, 1);
+    }
+
+    // Create the second project with second GitHub mock instance.
+    GitHubProject secondProject = new GitHubProject("test", "project2");
+    Path project2Dir = directoryFor(secondProject);
+    try (Repository repository =
+        FileRepositoryBuilder.create(project2Dir.resolve(".git").toFile())) {
+      repository.create();
+
+      fetcher = new TestGitHubDataFetcher(github);
+      fetcher.addRepositoryInfoForTesting(secondProject, project2Dir);
+      testLocalRepositoryFor(secondProject, 2);
+    }
+  }
+
+  @Test
+  public void testLocalRepsitoriesCapacity() throws IOException {
+    GitHub github = mock(GitHub.class);
+    fetcher = new TestGitHubDataFetcher(github);
+    for (int i = 0; i < 110; i++) {
+      fetcher.addForTesting(mock(GitHubProject.class), mock(LocalRepository.class));
+    }
+    assertEquals(LOCAL_REPOSITORIES_CACHE_CAPACITY, LOCAL_REPOSITORIES.size());
```

Comment:
Again, does it make sense to make sure that the content of the cache is correct? Or not?

Reply:
Yes, it does. I have implemented that.